packages/core/profiler/src/Tracer.js

```diff
@@ -64,44 +65,44 @@ export default class Tracer {
     return this.#traceEmitter.addListener(cb);
   }
 
-  async wrap(name: string, fn: () => mixed): Promise<void> {
-    let measurement = this.createMeasurement(name);
-    try {
-      await fn();
-    } finally {
-      measurement && measurement.end();
+  measure<T>(
+    {args = {}, categories, name}: MeasurementOptions,
+    fn: () => T,
+  ): T {
+    if (!this.enabled) {
+      return fn();
     }
-  }
 
-  createMeasurement(
```

Comment:
The original reasoning for having an API that didn't involve passing an object was to avoid the overhead of unnecessary object allocation in the case when tracing was disabled (which is most of the time).

Have you checked that the impact of performance on builds with tracing disabled of this change? Does timing/memory usage look okay? I just want to make sure we're not introducing unnecessary GC thrashing that might affect builds the size of Jira's.

Reply:
I created a crude benchmark that can be run with `node --trace-gc --trace-gc_verbose test.mjs` with this, commenting in one test at a time. I will revert the existing change so that it uses something more performant with more flexibility.

```js
let enabled = false;

class Measurement {
  constructor(args) {
    this.args = args;
    this.start = performance.now();
  }

  end() {
    const end = performance.now();

    // tracer.trace();
  }
}

async function measure(args, fn) {
  if (!enabled) {
    return fn();
  }

  const measurement = new Measurement(args);
  let result;
  try {
    result = await fn();
  } finally {
    measurement.end();
  }

  return result;
}

for (let i = 0; i < 100_000_000; i++) {
  // fastest -> slowest
  await measure('world', async () => {});
  // await measure(enabled && { hello: 'world' }, async () => {});
  // await measure(enabled ? { hello: 'world' } : null, async () => {});
  // await measure({ hello: 'world' }, async () => {});
  // await measure(() => ({ hello: 'world' }), async () => {});
  // await measure({ get hello() { return 'world' }}, async () => {});
}

// for (let i = 0; i < 1_000_000_000; i++) {
//   measure({ hello: 'world' }, () => {});
//   measure('world', () => {});
// }
```